codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -496,19 +530,52 @@ private void incrementPendingBytes(int numBytes) {
              */
             int write(int allowedBytes) {
                 int before = payload.size();
-                needFlush |= payload.write(Math.max(0, allowedBytes));
-                int writtenBytes = before - payload.size();
+                int writtenBytes = 0;
+                try {
+                    if (writing) {
+                        throw new IllegalStateException("write is not re-entrant");
+                    }
+                    // Write the portion of the frame.
+                    writing = true;
+                    needFlush |= payload.write(Math.max(0, allowedBytes));
+                    if (!cancelled && payload.size() == 0) {
+                        // This frame has been fully written, remove this frame
+                        // and notify the payload. Since we remove this frame
+                        // first, we're guaranteed that its error method will not
+                        // be called when we call cancel.
+                        pendingWriteQueue.remove();
+                        payload.writeComplete();
+                    }
+                } catch (Throwable e) {
+                    // Mark the state as cancelled, we'll clear the pending queue
+                    // via cancel() below.
+                    cancelled = true;
```

Comment:
@buchgr - I agree about pulling the code apart would be good to avoid if we can.  I would just like to make sure that the original exception trace is not lost here.  For example what about having `cancel` take a `Throwable` argument (or add another internal `cancel` method with a parameter)?  Just some way to have the [streamError](https://github.com/netty/netty/blob/master/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java#L465) call propagate the exception (i.e. use the [other streamError](https://github.com/netty/netty/blob/master/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Exception.java#L106) method with the `Throwable` argument).